File: 3. Behavioral Design Patterns/Memento/Java/MementoTemplate.java
// Memento Pattern – Java Template

import java.util.*;

class Memento {
    private final String state;
    public Memento(String state) { this.state = state; }
    public String getState() { return state; }
}

class Originator {
    private String state;
    public void setState(String state) { this.state = state; }
    public String getState() { return state; }
    public Memento createMemento() { return new Memento(state); }
    public void restoreMemento(Memento memento) { state = memento.getState(); }
}

class Caretaker {
    private final Stack<Memento> mementos = new Stack<>();
    public void saveState(Originator originator) { mementos.push(originator.createMemento()); }
    public void restoreState(Originator originator) { if (!mementos.isEmpty()) originator.restoreMemento(mementos.pop()); }
}

public class MementoTemplate {
    public static void main(String[] args) {
        Originator originator = new Originator();
        Caretaker caretaker = new Caretaker();
        originator.setState("State1");
        caretaker.saveState(originator);
        originator.setState("State2");
        caretaker.restoreState(originator);
    }
}

File: 3. Behavioral Design Patterns/Memento/C#/MementoTemplate.cs
// Memento Pattern – C# Template

using System.Collections.Generic;

namespace MementoTemplate
{
    class Memento
    {
        public string State { get; }
        public Memento(string state) { State = state; }
    }

    class Originator
    {
        private string state;
        public void SetState(string state) { this.state = state; }
        public string GetState() => state;
        public Memento CreateMemento() => new Memento(state);
        public void RestoreMemento(Memento memento) { state = memento.State; }
    }

    class Caretaker
    {
        private readonly Stack<Memento> mementos = new();
        public void SaveState(Originator originator) { mementos.Push(originator.CreateMemento()); }
        public void RestoreState(Originator originator) { if (mementos.Count > 0) originator.RestoreMemento(mementos.Pop()); }
    }

    class Program
    {
        static void Main()
        {
            var originator = new Originator();
            var caretaker = new Caretaker();
            originator.SetState("State1");
            caretaker.SaveState(originator);
            originator.SetState("State2");
            caretaker.RestoreState(originator);
        }
    }
}
